File: apps/core/src/main/java/dev/audiobook/platform/status/PlatformStatus.java
package dev.audiobook.platform.status;

public record PlatformStatus(String apiVersion, Build build, AvailabilityState availability) {

    public PlatformStatus(
            String apiVersion,
            String buildVersion,
            String buildRevision,
            Availability coreAvailability,
            Availability databaseAvailability) {
        this(
                apiVersion,
                new Build(buildVersion, buildRevision),
                new AvailabilityState(coreAvailability, databaseAvailability));
    }

    public record Build(String version, String revision) {
    }

    public record AvailabilityState(Availability core, Availability database) {
    }

    public enum Availability {
        AVAILABLE,
        DEGRADED
    }
}

File: apps/web/src/public-studio/platform-status.ts
export type Availability = "AVAILABLE" | "DEGRADED";

export interface PlatformStatus {
  apiVersion: string;
  build: {
    version: string;
    revision: string;
  };
  availability: {
    core: Availability;
    database: Availability;
  };
}

export async function fetchPlatformStatus(signal: AbortSignal): Promise<PlatformStatus> {
  const response = await fetch("/api/v1/platform/status", {
    headers: { Accept: "application/json" },
    signal
  });

  if (!response.ok) {
    throw new Error(`Platform status returned ${response.status}`);
  }

  return response.json() as Promise<PlatformStatus>;
}
